app/src/main/java/chat/rocket/android/authentication/login/presentation/LoginPresenter.kt

```diff
@@ -438,54 +497,162 @@ class LoginPresenter @Inject constructor(
-            it["clientId"] ?: it["appId"]
-        }.toString()
+    /**
+     * Returns an OAuth service map given a [serviceName].
+     *
+     * @param listMap The list of [Map] to get the service from.
+     * @param serviceName The service name to get in the [listMap]
+     * @return The OAuth service map or null otherwise.
+     */
+    private fun getServiceMap(
+        listMap: List<Map<String, Any>>,
+        serviceName: String
+    ): Map<String, Any>? {
+        return listMap.find { map -> map.containsValue(serviceName) }
     }
 
-    private fun getSamlServices(listMap: List<Map<String, Any>>): List<Map<String, Any>> {
-        return listMap.filter { map -> map["service"] == "saml" }
+    /**
+     * Returns the OAuth client ID of a [serviceMap].
+     * REMARK: This function works for common OAuth providers (Google, Facebook, Github and so on)
+     * as well as custom OAuth.
+     *
+     * @param serviceMap The service map to get the OAuth client ID.
+     * @return The OAuth client ID or null otherwise.
+     */
+    private fun getOauthClientId(serviceMap: Map<String, Any>): String? {
+        return try {
+            serviceMap["clientId"] as String ?: serviceMap["appId"] as String
+        } catch (exception: ClassCastException) {
+            null
+        }
     }
 
-    private fun getSamlServiceName(service: Map<String, Any>): String {
-        return service["buttonLabelText"].toString()
+    /**
+     * Returns a custom OAuth service list.
+     *
+     * @return A custom OAuth service list, otherwise an empty list if there is no custom OAuth service.
+     */
+    private fun getCustomOauthServices(listMap: List<Map<String, Any>>): List<Map<String, Any>> {
+        return listMap.filter { map -> map["custom"] == true }
     }
 
-    private fun getSamlProvider(service: Map<String, Any>): String {
-        return (service["clientConfig"] as Map<*, *>)["provider"].toString()
+    /** Returns the custom OAuth service host.
+     *
+     * @param serviceMap The service map to get the custom OAuth service host.
+     * @return The custom OAuth service host, otherwise null.
+     */
+    private fun getCustomOauthHost(serviceMap: Map<String, Any>): String? {
+        return try {
+            serviceMap["serverURL"] as String
+        } catch (exception: ClassCastException) {
+            null
+        }
     }
 
-    private fun getCustomOauthServices(listMap: List<Map<String, Any>>): List<Map<String, Any>> {
-        return listMap.filter { map -> map["custom"] == true }
+    /** Returns the custom OAuth service authorize path.
+     *
+     * @param serviceMap The service map to get the custom OAuth service authorize path.
+     * @return The custom OAuth service authorize path, otherwise null.
+     */
+    private fun getCustomOauthAuthorizePath(serviceMap: Map<String, Any>): String? {
+        return try {
+            serviceMap["authorizePath"] as String
+        } catch (exception: ClassCastException) {
```

Comment:
You can use `serviceMap["authorizePath"] as? String` instead... https://kotlinlang.org/docs/reference/typecasts.html#safe-nullable-cast-operator